.github/workflows/test.yaml

```
@@ -38,8 +38,18 @@ jobs:
         - functionality/memory
         - functionality/planner
         - functionality/hooks
+        - test_controller
         
     steps:
       - uses: actions/checkout@v4
       - uses: astral-sh/setup-uv@v5
-      - run: uv run --with=dotenv pytest tests/${{ matrix.test }}.py || true
+      - name: Setup virtual environment and install dependencies
+        if: matrix.test == 'test_controller'
+        run: |
+          uv venv
+          uv pip install -e .
```

```suggestion
          uv sync
```

we may want to use sync instead so it installs the dev packages used for tests